File: ProgAsynchrone-ChatApp/ClientsPool/C#/Server/ClientHandler.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace Server
{
    internal class ClientHandler
    {
        private readonly TcpClient _client;
        private StreamWriter _writer;

        public ClientHandler(TcpClient client)
        {
            _client = client;
        }

        public async Task HandleClientMessagesAsync()
        {
            using (var stream = _client.GetStream())
            {
                using (var reader = new StreamReader(stream))
                {
                    _writer = new StreamWriter(stream) { AutoFlush = true };

                    // Read user name
                    await _writer.WriteLineAsync("Enter your name: ");
                    string userName = await reader.ReadLineAsync();

                    string message = userName + " has joined the chat.";
                    Console.WriteLine(message);
                    this.Broadcast(message);

                    try
                    {
                        string receivedMessage;
                        while ((receivedMessage = await reader.ReadLineAsync()) != null)
                        {
                            message = userName + ": " + receivedMessage;
                            Console.WriteLine(message);
                            Broadcast(message);
                        }
                    }
                    catch (IOException)
                    {
                        // Handle disconnection
                    }
                    finally
                    {
                        Console.WriteLine($"{userName} has left the chat.");
                        Broadcast($"{userName} has left the chat.");
                    }
                }
            };
        }

        private void Broadcast(string message)
        {
            Program.Broadcast(message, this);
        }

        public void SendMessage(string message)
        {
            if (_writer != null)
            {
                _writer.WriteLine(message);
            }
        }
    }
}

File: ProgAsynchrone-ChatApp/ClientsPool/Java/Server/src/ClientHandler.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.Socket;

public class ClientHandler implements Runnable {
    private final Socket clientSocket;
    private PrintWriter writer;

    public ClientHandler(Socket clientSocket) {
        this.clientSocket = clientSocket;
    }

    @Override
    public void run() {
        try (var stream = clientSocket.getInputStream();
             var reader = new BufferedReader(new InputStreamReader(stream));) {

            writer = new PrintWriter(new OutputStreamWriter(clientSocket.getOutputStream()), true);

            // Read user name
            writer.println("Enter your name: ");
            String userName = reader.readLine();
            String message = userName + " has joined the chat.";
            System.out.println(message);
            broadcast(message);

            try {
                String messageReceived;
                while ((messageReceived = reader.readLine()) != null) {
                    message = userName + ": " + messageReceived;
                    System.out.println(message);
                    broadcast(message);
                }
            } catch (IOException e) {
                // Handle disconnection
            } finally {
                System.out.println(userName + " has left the chat.");
                broadcast(userName + " has left the chat.");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void broadcast(String message){
        Main.broadcast(this, message);
    }

    public void sendMessage(String message) {
        writer.println(message);
    }
}